apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/sub-task-agent.ts

```diff
@@ -0,0 +1,725 @@
+import type { CompletedToolCall } from "./types";
+import { mapPreviousCallsToLlmMessages } from "./util";
+
+const model: LlmParams["model"] = "claude-3-opus-20240229";
+
+const omittedCoordinatorToolNames = [
+  "complete",
+  "startFactGatheringSubTasks",
+  "proposeEntitiesFromFacts",
+  "requestHumanInput",
+  "submitProposedEntities",
+  "terminate",
+] as const;
+
+type OmittedCoordinatorToolNames = Subtype<
+  CoordinatorToolName,
+  (typeof omittedCoordinatorToolNames)[number]
+>;
+
+const subTaskAgentCustomToolNames = ["complete", "terminate"] as const;
+
+type SubTaskAgentCustomToolName = (typeof subTaskAgentCustomToolNames)[number];
+
+type SubTaskAgentToolName =
+  | OmitValue<CoordinatorToolName, OmittedCoordinatorToolNames>
+  | SubTaskAgentCustomToolName;
+
+const generateToolDefinitions = <
+  T extends SubTaskAgentCustomToolName[],
+>(params: {
+  omitTools: T;
+}): Record<
+  OmitValue<SubTaskAgentToolName, T[number]>,
+  LlmToolDefinition<OmitValue<SubTaskAgentToolName, T[number]>>
+> => {
+  const coordinatorToolDefinitions = generateCoordinatorToolDefinitions({
+    omitTools: omittedCoordinatorToolNames.concat(),
+  });
+
+  const allToolDefinitions = {
+    ...coordinatorToolDefinitions,
+    terminate: {
+      name: "terminate",
+      description:
+        "Terminate the sub-task, because you cannot find the required information to complete it.",
+      inputSchema: {
+        type: "object",
+        properties: {
+          reason: {
+            type: "string",
+            description: "The reason for terminating the sub-task.",
+          },
+        },
+        required: ["reason"],
+      },
+    },
+    complete: {
+      name: "complete",
+      description: "Complete the sub-task.",
+      inputSchema: {
+        type: "object",
+        properties: {
+          explanation: {
+            type: "string",
+            description:
+              "The explanation for how the gathered facts satisfy sub-task.",
+          },
+        },
+      },
+    },
+  };
+
+  const filteredToolDefinitions = Object.fromEntries(
+    Object.entries(allToolDefinitions).filter(
+      ([toolName]) => !params.omitTools.includes(toolName as T[number]),
+    ),
+  ) as Record<
+    OmitValue<SubTaskAgentToolName, T[number]>,
+    LlmToolDefinition<OmitValue<SubTaskAgentToolName, T[number]>>
+  >;
+
+  return filteredToolDefinitions;
+};
+
+export type SubTaskAgentToolCallArguments = Subtype<
+  Record<SubTaskAgentToolName, unknown>,
+  {
+    complete: {
+      explanation: string;
+    };
+    terminate: {
+      reason: string;
+    };
+  } & Omit<CoordinatorToolCallArguments, OmittedCoordinatorToolNames>
+>;
+
+const generateSystemPromptPrefix = (params: { input: SubTaskAgentInput }) => {
+  const {
+    relevantEntities,
+    existingFactsAboutRelevantEntities,
+    linkEntityTypes,
+  } = params.input;
+
+  return dedent(`
+    You are a sub-task agent for a research task.
+
+    The user will provide you with:
```

Comment:
Does this agent need to know it is a 'sub-task' agent? It otherwise looks like it's completing a research goal the same as any other, and we only mention that it's a sub-task of some wider goal here.